src/abstractdataframe/abstractdataframe.jl

```diff
@@ -67,6 +67,73 @@ _names(df::AbstractDataFrame) = _names(index(df))
 
 Compat.hasproperty(df::AbstractDataFrame, s::Symbol) = haskey(index(df), s)
 
+"""
+    rename!(df::AbstractDataFrame, vals::AbstractVector{Symbol}; makeunique::Bool=false)
+    rename!(df::AbstractDataFrame, vals::AbstractVector{AbstractString}; makeunique::Bool=false)
+    rename!(df::AbstractDataFrame, (from => to)::Pair...)
+    rename!(df::AbstractDataFrame, d::AbstractDict)
+    rename!(df::AbstractDataFrame, d::AbstractArray{Pair})
+    rename!(f::Function, df::AbstractDataFrame)
+
+Rename columns of `df` in-place.
+Each name is changed at most once. Permutation of names is allowed.
+
+# Arguments
+- `df` : the `AbstractDataFrame`
+- `d` : an `AbstractDict` type or an `AbstractVector` of `Pairs` that maps
+  the original names or column numbers to new names
+- `f` : a function which for each column takes the old name (a `Symbol`)
+  and returns the new name that gets converted to a `Symbol`
+- `vals` : new column names as a vector of `Symbol`s or `AbstractStrings`
+  of the same length as the number of columns in `df`
+- `makeunique` : if `false` (the default), an error will be raised
+  if duplicate names are found; if `true`, duplicate names will be suffixed
+  with `_i` (`i` starting at 1 for the first duplicate).
+
+If pairs are passed to `rename!` (as positional arguments or by a dictionary or a vector)
+then:
+* `from` value can be a `Symbol`, an `AbstractString` or an `Integer`;
+* `to` value can be a `Symbol`, an `AbstractString`.
+
+It is required that the type of `from` and of `to` values is the same (e.g.
+it is not allowed to mix strings and symbols as `from` values).
+
+See also: [`rename`](@ref)
+
+# Examples
+```julia
+julia> df = DataFrame(i = 1, x = 2, y = 3)
+1×3 DataFrame
+│ Row │ i     │ x     │ y     │
+│     │ Int64 │ Int64 │ Int64 │
+├─────┼───────┼───────┼───────┤
+│ 1   │ 1     │ 2     │ 3     │
+
+julia> rename!(df, Dict(:i => "A", :x => "X"))
+1×3 DataFrame
+│ Row │ A     │ X     │ y     │
+│     │ Int64 │ Int64 │ Int64 │
+├─────┼───────┼───────┼───────┤
+│ 1   │ 1     │ 2     │ 3     │
+
+julia> rename!(df, [:a, :b, :c])
+1×3 DataFrame
+│ Row │ a     │ b     │ c     │
+│     │ Int64 │ Int64 │ Int64 │
+├─────┼───────┼───────┼───────┤
+│ 1   │ 1     │ 2     │ 3     │
+
+julia> rename!(df, [:a, :b, :a])
+ERROR: ArgumentError: Duplicate variable names: :a. Pass makeunique=true to make them unique using a suffix automatically.
+
+julia> rename(df, [:a, :b, :a], makeunique=true)
+1×3 DataFrame
+│ Row │ a     │ b     │ a_1   │
+│     │ Int64 │ Int64 │ Int64 │
+├─────┼───────┼───────┼───────┤
+│ 1   │ 1     │ 2     │ 3     │
+```
```

Comment:
Add an example with a function while you're at it?

Reply:
added (I had it in `rename` but it is ok to add it here also)